webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/domain/policies/RuleTarget.scala

```diff
@@ -315,49 +318,36 @@ object RuleTarget extends Loggable {
           case AllTargetExceptPolicyServers => nodes ++ allNodes.collect { case (k, isPolicyServer) if (!isPolicyServer) => k }
           case AllPolicyServers             => nodes ++ allNodes.collect { case (k, isPolicyServer) if (isPolicyServer) => k }
           case PolicyServerTarget(nodeId)   =>
-            if (allNodesAreThere) {
               nodes :+ nodeId
-            } else {
-              // nodeId may not be in allNodes
-              // allNodes.keys.exists(x => x == nodeId)
-              Chunk.fromIterable(allNodes.keys).contains(nodeId) match {
-                case true => nodes :+ nodeId
-                case _    => nodes
-              }
-            }
 
           // here, if we don't find the group, we consider it's an error in the
           // target recording, but don't fail, just log it.
           case GroupTarget(groupId)         =>
             val groupNodes = groups.getOrElse(groupId, Chunk.empty)
             val filtered   = {
-              if (allNodesAreThere) groupNodes
-              else {
-                val keys = Chunk.fromIterable(allNodes.keys)
-                groupNodes.filter(keys.contains(_))
```

Comment:
I think this is the culprit : a `contains` on a `Chunk` is `O(n)`, it should have been `allNodes.contains(_)` on the Map which is `O(1)`